sdks/go/pkg/beam/runners/direct/direct.go

```diff
@@ -34,229 +34,233 @@ func Execute(ctx context.Context, p *beam.Pipeline) error {
-// methods on units?
-
-func getEdge(unit exec.Unit) (*graph.MultiEdge, bool) {
-	switch unit.(type) {
-	case *exec.ParDo:
-		return unit.(*exec.ParDo).Edge, true
-	case *exec.Combine:
-		return unit.(*exec.Combine).Edge, true
-	case *GBK:
-		return unit.(*GBK).Edge, true
-	case *exec.DataSource:
-		return unit.(*exec.DataSource).Edge, true
-	case *exec.DataSink:
-		return unit.(*exec.DataSink).Edge, true
-	case *Impulse:
-		return unit.(*Impulse).Edge, true
-	default:
-		return nil, false
+func (b *builder) makeLinks(ids []linkID) ([]exec.Node, error) {
+	var ret []exec.Node
+	for _, id := range ids {
+		n, err := b.makeLink(id)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, n)
 	}
+	return ret, nil
 }
 
-func setOut(unit exec.Unit, out []exec.Node) {
-	switch unit.(type) {
-	case *exec.ParDo:
-		unit.(*exec.ParDo).Out = out
-	case *exec.Combine:
-		unit.(*exec.Combine).Out = out
-	case *GBK:
-		if len(out) != 1 {
-			panic(fmt.Errorf("bad outputs for GBK: %v", out))
+func (b *builder) makeLink(id linkID) (exec.Node, error) {
+	if n, ok := b.links[id]; ok {
+		return n, nil
+	}
+
+	// Process all incoming links for the edge and cache them. It thus doesn't matter
+	// which exact link triggers the Node generation. The link caching is only needed
+	// to process ParDo side inputs.
+
+	edge := b.edges[id.to]
+
+	out, err := b.makeNodes(edge.Output)
+	if err != nil {
+		return nil, err
+	}
+
+	var u exec.Node
+	switch edge.Op {
+	case graph.ParDo:
+		pardo := &exec.ParDo{UID: b.idgen.New(), Edge: edge, Out: out}
+		if len(edge.Input) == 1 {
+			u = pardo
+			break
 		}
-		unit.(*GBK).Out = out[0]
-	case *exec.DataSource:
-		unit.(*exec.DataSource).Out = out[0]
-	case *exec.DataSink:
-		// nop
-	case *Impulse:
-		unit.(*Impulse).Out = out
+
+		// ParDo w/ side input. We need to insert buffering and wait. We also need to
+		// ensure that we return the correct link node.
+
+		b.units = append(b.units, pardo)
+
+		w := &wait{UID: b.idgen.New(), need: len(edge.Input) - 1, next: pardo}
+		b.units = append(b.units, w)
+		b.links[linkID{edge.ID(), 0}] = w
+
+		for i := 1; i < len(edge.Input); i++ {
+			n := &buffer{uid: b.idgen.New(), next: w.ID(), read: pardo.ID(), notify: w.notify}
+			pardo.Side = append(pardo.Side, n)
+
+			b.units = append(b.units, n)
+			b.links[linkID{edge.ID(), i}] = n
+		}
+
+		return b.links[id], nil
+
+	case graph.Combine:
+		u = &exec.Combine{UID: b.idgen.New(), Edge: edge, Out: out[0]}
+
+	case graph.GBK:
+		u = &GBK{UID: b.idgen.New(), Edge: edge, Out: out[0]}
+
+	case graph.Flatten:
+		u = &exec.Flatten{UID: b.idgen.New(), N: len(edge.Input), Out: out[0]}
+
+		for i := 0; i < len(edge.Input); i++ {
+			b.links[linkID{edge.ID(), i}] = u
+		}
+
+	case graph.DataSink:
+		u = &exec.DataSink{UID: b.idgen.New(), Edge: edge}
+
 	default:
-		panic(fmt.Sprintf("Unit %v has no output", unit))
+		return nil, fmt.Errorf("unexpected edge: %v", edge)
```

Comment:
Brainstorming about the beam.External issue, is it worth building an "unknown node" and linking it in here, and returning the unknown nodes as a return value of this function? That way, a caller could verify all the unknown nodes are beam.External nodes and be happy with the translation. As is, this code fails on the first translation error, and the caller won't be able to detect if there are other issues in the graph.

Reply:
I think the External edges should be resolved/expanded before or at the time we reach this point. Although we're using the Compile function in the harness as well, I was rather thinking it should be a private part of the direct runner (see BEAM-3316).